make/jdk/src/classes/build/tools/taglet/SealedGraph.java

```diff
@@ -0,0 +1,260 @@
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package build.tools.taglet;
+
+import com.sun.source.doctree.DocTree;
+import java.nio.file.Path;
+import jdk.javadoc.doclet.Doclet;
+import jdk.javadoc.doclet.DocletEnvironment;
+import jdk.javadoc.doclet.Taglet;
+
+import javax.lang.model.element.*;
+import javax.lang.model.type.DeclaredType;
+
+import java.io.*;
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static java.lang.System.lineSeparator;
+import static jdk.javadoc.doclet.Taglet.Location.TYPE;
+
+/**
+ * A block tag to optionally insert a reference to a sealed class hierarchy graph,
+ * and generate the corresponding dot file.
+ */
+public final class SealedGraph implements Taglet {
+    private static final String sealedDotOutputDir =
+            System.getProperty("sealedDotOutputDir");
+
+    private DocletEnvironment docletEnvironment;
+
+    /**
+     * Returns the set of locations in which a taglet may be used.
+     */
+    @Override
+    public Set<Location> getAllowedLocations() {
+        return EnumSet.of(TYPE);
+    }
+
+    @Override
+    public boolean isInlineTag() {
+        return false;
+    }
+
+    @Override
+    public String getName() {
+        return "sealedGraph";
+    }
+
+    @Override
+    public void init(DocletEnvironment env, Doclet doclet) {
+        docletEnvironment = env;
+    }
+
+    @Override
+    public String toString(List<? extends DocTree> tags, Element element) {
+        if (sealedDotOutputDir == null || sealedDotOutputDir.isEmpty()) {
+            return "";
+        }
+        if (docletEnvironment == null || !(element instanceof TypeElement typeElement)) {
+            return "";
+        }
+
+        ModuleElement module = docletEnvironment.getElementUtils().getModuleOf(element);
+        Path dotFile = Path.of(sealedDotOutputDir,
+                module.getQualifiedName() + "_" + typeElement.getQualifiedName() + ".dot");
+
+        Set<String> exports = module.getDirectives().stream()
+                .filter(ModuleElement.ExportsDirective.class::isInstance)
+                .map(ModuleElement.ExportsDirective.class::cast)
+                // Only include packages that are globally exported (i.e. no "to" exports)
+                .filter(ed -> ed.getTargetModules() == null)
+                .map(ModuleElement.ExportsDirective::getPackage)
+                .map(PackageElement::getQualifiedName)
+                .map(Objects::toString)
+                .collect(Collectors.toUnmodifiableSet());
+
+        String dotContent = Renderer.graph(typeElement, exports);
+
+        try (PrintWriter pw = new PrintWriter(dotFile.toFile())) {
+            pw.println(dotContent);
+        } catch (FileNotFoundException e) {
+            throw new RuntimeException(e);
+        }
```

Comment:
It's certainly unusual for a taglet `.toString` call to have this sort of side effect.   It would not the a showstopper, but have you checked whether this code is just executed once or multiple times?

Reply:
This is Per's code; I didn't look in detail in it. It does seem to be the correct place, though; it's more like the name of the method is bad, since it overloads the `toString` from `Object`, and gives the wrong impressions of its meaning and usage.

However, it seems we need change some code to be guaranteed by called only once. Looking into it...

Reply:
Actually, no, it's fine. According to `toString` documentation:

```
     * If this taglet supports {@link #isBlockTag block} tags, it will be called once
     * for each comment containing instances of block tags, with a list of all the instances
     * of the block tag in that comment.
```

and I just presume we can trust the JDK authors to only add the tag once per class.